src/alpaca_typer.erl

```diff
@@ -1009,9 +1013,49 @@ inst_type_arrow(EnvIn, {type_constructor, Line, Name}) ->
                      Arrow = {type_arrow, inst_constructor_arg(Arg, Vs, Types), ADT},
                      {Env, Arrow}
              end,
-    Default = {error, {bad_constructor, Line, Name}},
-    C = proplists:get_value(Name, EnvIn#env.type_constructors, Default),
-    try Cons_f(ADT_f(C))
+
+    %% If the constructor was not qualified with a module name it's pretty
+    %% to fetch but if it was, then we need to first make sure it's in
+    %% the specified module *and* that its enclosing type is exported.
+    %%
+    %% Here's the easy local get:
+    GetTC = fun(#type_constructor{line=Line, name=Name, module=undefined}) ->
+                    Default = {error, {bad_constructor, Line, Name}},
+                    %% constructors defined in this module or imported by it:
+                    Available = EnvIn#env.type_constructors,
+                    C = proplists:get_value(Name, Available, Default);
+
+               %% and the part where we go to a different module:
+               (#type_constructor{line=Line, name=Name, module=Mod}) ->
+                    Mods = EnvIn#env.modules,
+                    M = [AM || #alpaca_module{name=N}=AM <- Mods, Mod =:= N],
+                    case M of
+                        [] ->
+                            throw({error, {bad_module, Line, Mod}});
+                        [Target] ->
+                            %% in the beginning of typing, constructors/1 links
+                            %% the actual type to the constructor contained
+                            %% within it to make it easy for inst_type to
+                            %% instantiate the type itself.  Here we grab each
+                            %% constructor whose name matches the one we're
+                            %% looking for and insert the ADT itself in a
+                            %% similar manner.
+                            Types = Target#alpaca_module.types,
+                            F = fun(#alpaca_type{members=Ms}=AT) ->
+                                        [M#alpaca_constructor{type=AT} ||
+                                            #alpaca_constructor{
+                                               name=#type_constructor{name=TCN}
+                                              }=M <- Ms, TCN =:= Name]
+                                     end,
+                            case lists:flatten(lists:map(F, Types)) of
+                                [RealC|_] ->
```

Comment:
Good point, the AST gen state does ensure unique constructors within a module.  I was being overly cautious I think :)